syntax/encoding/yaml.go

```diff
@@ -54,6 +63,22 @@ type linePosition struct {
 
 type positionIndex struct {
 	lines []linePosition
+	path  []string
+}
+
+const indexRegexPattern = `\[\d+\]`
+
+func (p positionIndex) pathString() string {
+	var pathString string
+	for _, s := range p.path {
+		re := regexp.MustCompile(indexRegexPattern)
+		if re.MatchString(s) {
+			pathString += s
+			continue
+		}
+		pathString = util.JoinKey(pathString, s)
+	}
+	return pathString
```

Comment:
I think we should keep path components in their typed representations here. That would allow us to match on types rather than using regexes.

```suggestion
	path  []any
}

func (p positionIndex) pathString() string {
	var pathString string
	for _, s := range p.path {
		switch s := s.(type) {
		case int:
			pathString = fmt.Sprintf("%s[%d]", pathString, s)
		case string:
			pathString = util.JoinKey(pathString, s)
		}
	}
	return pathString
```

Reply:
👍🏽 done